### PR TITLE
Hardcoded immutable samplers in Metal

### LINKED ISSUE #2837: Hard-code immutable samplers into MSL
Some iOS devices don't support comparison samplers being bound dynamically at all. Instead one can hard-code them in the generated MSL, which SPIRV-Cross can already do, see https://github.com/KhronosGroup/MoltenVK/pull/644

### PR DESCRIPTION
**Pull request by kvark**
Also adds unnormalized sampling to HAL.
Blocked on spirv_cross 0.15.0 release.

Fixes #2837
Partially fulfills #2883 
PR checklist:
- [x] `make` succeeds (on *nix)
- [x] `make reftests` succeeds
- [x] tested examples with the following backends: metal
- [ ] `rustfmt` run on changed code

### CODE REVIEW

**Comment by kvark** [comment]
fwiw, I confirmed on VK CTS that the samplers get hard-coded and work as expected.

File: src/backend/metal/src/device.rs [hunk 1]
```diff
@@ -795,18 +809,67 @@ impl Device {
...
+            min_filter: match info.min_filter {
+                image::Filter::Nearest => msl::SamplerFilter::Nearest,
+                image::Filter::Linear => msl::SamplerFilter::Linear,
+            },
+            mag_filter: match info.mag_filter {
+                image::Filter::Nearest => msl::SamplerFilter::Nearest,
+                image::Filter::Linear => msl::SamplerFilter::Linear,
+            },
+            mip_filter: match info.min_filter {
+                image::Filter::Nearest if info.lod_range.end < image::Lod::from(0.5) =>
```
Comment (grovesNL) [inline src/backend/metal/src/device.rs]: Why do we need the special LOD logic here?

File: src/backend/vulkan/src/device.rs [hunk 2]
```diff
@@ -1096,7 +1096,7 @@ impl d::Device<B> for Device {
                     vk::BorderColor::FLOAT_TRANSPARENT_BLACK
                 }
             },
-            unnormalized_coordinates: vk::FALSE,
+            unnormalized_coordinates: if sampler_info.normalized { vk::TRUE } else { vk::FALSE },
```
[the remark this answers is not in the document]
Comment (kvark) [reply src/backend/vulkan/src/device.rs]: well spotted!

File: src/backend/metal/src/device.rs (hunk 1 above)
Comment (kvark) [reply src/backend/metal/src/device.rs]: same logic as we do in `make_sampler_descriptor`

**Comment by kvark** [comment]
Thanks for the review @grovesNL !
comments are addressed
bors r=grovesNL

**Comment by kvark** [comment]
bors retry

**Pull request merged**